bioblend/galaxy/dataset_collections/__init__.py

```diff
@@ -80,3 +94,109 @@ def __init__(self, name, id):
+            'id': dataset_collection_id,
+            'instance_type': instance_type,
+        }
+        url = self._make_url(module_id=dataset_collection_id)
+        return self._get(url=url, params=params)
+
+    def download_dataset_collection(self, dataset_collection_id: str, file_path: str,
+                                    history_id: Optional[str] = None):
+        """
+        Download a history dataset collection as a tgz archive.
+
+        :type dataset_collection_id: str
+        :param dataset_collection_id: Encoded dataset collection ID
+
+        :type file_path: str
+        :param file_path: The path to which the archive will be downloaded
+
+        :type history_id: str
+        :param history_id: Encoded ID of the collection's history
+
+        .. note::
+          This method is only supported by Galaxy 18.01 or later.
+        """
+        url = self._make_url(module_id=dataset_collection_id) + '/download'
+        params = {'history_id': history_id} if history_id else None
+        r = self.gi.make_get_request(url, stream=True, params=params)
+        r.raise_for_status()
+
+        with open(file_path, 'wb') as fp:
+            for chunk in r.iter_content(chunk_size=CHUNK_SIZE):
+                if chunk:
+                    fp.write(chunk)
+
+    def wait_for_dataset_collection(self, dataset_collection_id: str, maxwait: float = 12000,
+                                    interval: float = 3, proportion_complete: float = 1.0,
+                                    check: bool = True) -> dict:
+        """
+        Wait until a dataset is in a terminal state.
+
+        :type dataset_id: str
+        :param dataset_id: dataset ID
+
+        :type maxwait: float
+        :param maxwait: Total time (in seconds) to wait for the dataset state to
+          become terminal. If the dataset state is not terminal within this
+          time, a ``DatasetTimeoutException`` will be raised.
+
+        :type interval: float
+        :param interval: Time (in seconds) to wait between 2 consecutive checks.
+
+        :type check: bool
+        :param check: Whether to check if all datasets terminal states are 'ok'.
+
+        :rtype: dict
+        :return: Details of the given dataset collection.
+        """
+        assert maxwait >= 0
+        assert interval > 0
+        assert 0 <= proportion_complete <= 1
+
+        time_left = maxwait
+        while True:
+            dataset_collection = self.gi.dataset_collections.show_dataset_collection(dataset_collection_id)
+            states = [elem['object']['state'] for elem in dataset_collection['elements']]
+            count = 0
+            for i, state in enumerate(states):
+                if state in TERMINAL_STATES:
+                    if check and state != 'ok':
+                        raise Exception(f"Dataset {dataset_collection['elements']['object'][i]['id']} is in terminal state {state}")
+                    count += 1
+            proportion = count / len(states)
```

Comment:
```suggestion
            terminal_states = [state for state in states if state in TERMINAL_STATES]
            if set(teminal_states) not in [{'ok'}, set()]:
                raise Exception(f"Dataset collection {dataset_collection_id} contains elements in terminal states: {', '.join(set(terminal_states)}")
            proportion = len(terminal_states) / len(states)
```

I'm not a big fan of that for loop - maybe something like this?